### ISSUE TITLE
Issue with 'util/perl/OpenSSL/paramnames.pm'

### ISSUE THREAD

**Issue by gvanem** [openssl/openssl#21152]
I made a comment [here](https://github.com/openssl/openssl/issues/21112#issuecomment-1580830082) regarding the following build error with my Strawberry Perl;
 `F:\util\StrawberryPerl\perl\bin\perl.exe -v`:
```
This is perl 5, version 18, subversion 4 (v5.18.4) built for MSWin32-x86-multi-thread-64int
```

I was encouraged to create an issue. So here goes a minimal example. 
I created an empty `<OPENSSL_ROOT>\gv-build` directory and `cd`'d into it and did:
```
F:\util\StrawberryPerl\perl\bin\perl.exe ../Configure VC-WIN64A-ONECORE no-brotli no-zlib no-zstd enable-quic 
``` 

Then invoking `nmake`, this error occurs in the first recipie (edited for clarity):
``` 
"F:\util\StrawberryPerl\perl\bin\perl.exe" "-I." -Mconfigdata "..\util\dofile.pl" "-omakefile" \
"..\crypto\params_idx.c.in" > crypto\params_idx.c

"my" variable $trieref masks earlier declaration in same scope at F:/MinGW32/src/inet/Crypto/OpenSSL/util/perl/OpenSSL/paramnames.pm line 640.

syntax error at F:/MinGW32/src/inet/Crypto/OpenSSL/util/perl/OpenSSL/paramnames.pm line 582, near "$cursor{"

Global symbol "$cursor" requires explicit package name at 
F:/MinGW32/src/inet/Crypto/OpenSSL/util/perl/OpenSSL/paramnames.pm line 584.

Global symbol "%trie" requires explicit package name at 
F:/MinGW32/src/inet/Crypto/OpenSSL/util/perl/OpenSSL/paramnames.pm line 588.

syntax error at F:/MinGW32/src/inet/Crypto/OpenSSL/util/perl/OpenSSL/paramnames.pm line 589, near "}"
syntax error at F:/MinGW32/src/inet/Crypto/OpenSSL/util/perl/OpenSSL/paramnames.pm line 636, near "}"
syntax error at F:/MinGW32/src/inet/Crypto/OpenSSL/util/perl/OpenSSL/paramnames.pm line 662, near "}"
syntax error at F:/MinGW32/src/inet/Crypto/OpenSSL/util/perl/OpenSSL/paramnames.pm line 673, near "}"
Compilation failed in require at ..\crypto\params_idx.c.in line 12.

BEGIN failed--compilation aborted at ..\crypto\params_idx.c.in line 12.
NMAKE : fatal error U1077: '"F:\util\StrawberryPerl\perl\bin\perl.exe" "-I." -Mconfigdata "..\util\dofile.pl" 
"-omakefile" "..\crypto\params_idx.c.in" > crypto\params_idx.c' : return code '0x1'
Stop.
``` 

I fail to understand this Perl issue. What is `%$cursor{$c}`?

The configure+build works fine with Cygwin's perl for targeting `djgpp` for example.
Perhaps it's time to upgrade to a newer Strawberry Perl?

**Comment by levitte** [comment]
It may be that some shortcuts were taken, that work with a very current perl but not quite in 5.18.  I'll have a look...

**Comment by levitte** [comment]
Hmmmm, may I ask you to update your checked out code?  The messages you show do not correspond to the current paramnames.pm as far as I can see.

**Comment by gvanem** [comment]
>  The messages you show do not correspond to the current paramnames.pm as far as I can see.

Really? 
* line 582: https://github.com/openssl/openssl/blob/master/util/perl/OpenSSL/paramnames.pm#L582.
   `$cursor = %$cursor{$c};`
* line 585: https://github.com/openssl/openssl/blob/master/util/perl/OpenSSL/paramnames.pm#L584.
  `$$cursor{'val'} = $name;`
* line 588: https://github.com/openssl/openssl/blob/master/util/perl/OpenSSL/paramnames.pm#L588.
  `return %trie;`

Some missing `my` stuff?

**Comment by gvanem** [comment]
I'd whish but the latest (?) change to `Configure` broke my attempt:
```
md gv-build
cd gv-build
F:\util\StrawberryPerl\perl\bin\perl.exe ../Configure VC-WIN64A-ONECORE no-brotli no-zlib no-zstd enable-quic
Configuring OpenSSL version 3.2.0-dev for target VC-WIN64A-ONECORE
Using os-specific seed configuration

Failure!  makefile wasn't produced.
Please read INSTALL.md and associated NOTES-* files.  You may also have to
look over your available compiler tool chain or change your configuration.

mkdir F:\MinGW32\src\inet\Crypto\OpenSSL\util\perl|OpenSSL: Invalid argument; Filnavnet, mappenavnet eller volumnavnesyntaksen er feil at ../Configure line 1912.
```
Inserting some debug:
```diff
--- a/Configure 2023-06-08 09:29:22
+++ b/Configure 2023-06-08 10:25:56
@@ -1909,6 +1909,7 @@
         my $f = basename($file);

         # Make sure the directories we're building in exists
+        print STDERR "DEBUG: d=$d\n";
         mkpath($d);

         my $res = abs2rel(catfile(absolutedir($d), $f), rel2abs($relativeto));
```

I see it's trying to create illegal path-names:
```
...
DEBUG: d=F:\MinGW32\src\inet\Crypto\OpenSSL\crypto
DEBUG: d=F:\MinGW32\src\inet\Crypto\OpenSSL\util\perl|OpenSSL

Failure!  makefile wasn't produced.
```
Where does this `|` come from?

**Comment by gvanem** [comment]
I spoke too soon. After some time, the build stopped at:
```perl
"F:\util\StrawberryPerl\perl\bin\perl.exe" "-I." "-Mconfigdata" "..\util\dofile.pl" "-omakefile" "..\providers\common\include\prov\der_digests.h.in" > providers\common\include\prov\der_digests.h

Undefined subroutine &oids_to_c::process_leaves called at ..\providers\common\include\prov\der_digests.h.in line 16.
NMAKE : fatal error U1077: '"F:\util\StrawberryPerl\perl\bin\perl.exe" "-I." "-Mconfigdata" "..\util\dofile.pl" "
-omakefile" "..\providers\common\include\prov\der_digests.h.in" > providers\common\include\prov\der_digests.h' : 
return code '0x1'
Stop.
```

Perhaps caused by my *n00b* attempt at fixing this `mkpath()` issue.

**Comment by levitte** [comment]
I'll create a PR for this, that'll make it easier to see what's going on.

**Comment by levitte** [comment]
As for that "mysterious `|`", it comes from #21117 (now merged)

**Comment by levitte** [comment]
paramnames.pm fixed in #21178.
